freqtrade/plugins/pairlist/MarketCapPairList.py

```diff
@@ -0,0 +1,263 @@
+                _id = self._coins_list[symbol]
+                # symbol is not unique (sigh!)
+                if type(_id) == list:
+                    pair = symbol.upper() + f'/{self._stake_currency}'
+                    _id_text = ', '.join(_id)
+                    self.log_once(f'Symbol for {pair} is not unique on coingecko '
+                                  f'({_id_text}), dropping.', logger.warning)
+                    continue
+                ids.append(_id)
+            elif symbol in self._coins_list.values():
+                # try to add some symbols automatically
+                symbol_map[symbol] = [k for k, v in self._coins_list.items() if v == symbol][0]
+                ids.append(symbol)
+
+        # reverse map
+        rev_symbol_map = {v: k for k, v in symbol_map.items()}
+
+        _marketcap_ranks = {}
+
+        # seems coingecko limits the number of ids to 52, using 50.
+        for _ids in chunks(ids, 50):
+            base_marketkaps = self._cg.get_coins_markets(
+                    vs_currency='usd',
+                    ids=','.join(_ids),
+                    order='market_cap_desc',
+                    per_page=len(ids),
+                    sparkline=False,
+                    page=1)
+            for x in base_marketkaps:
+                # same here, keep symbol lowercase.
+                _symbol = x['symbol'].lower()
+                if _symbol in rev_symbol_map:
+                    _symbol = rev_symbol_map[_symbol]
+                _marketcap_ranks[_symbol] = x['market_cap_rank']
+
+        for x in self._marketcap_ranks.keys():
+            if x not in _marketcap_ranks:
+                _marketcap_ranks[x] = None
+        self._marketcap_ranks = _marketcap_ranks
+
+    def update_marketcap_ranks(self, symbols: List[str]):
+        """
+        Updates the dict containing the marketcap ranks of the requested
+        list of symbols, if needed.
+        """
+        marketcaps_uptodate = self._marketcap_cache.get('marketcaps_uptodate')
+        startup = not (self._marketcap_ranks and self._coins_list)
+
+        for symbol in symbols:
+            if symbol.lower() not in self._marketcap_ranks:
+                marketcaps_uptodate = False
+                self._marketcap_ranks[symbol.lower()] = None
+        if not marketcaps_uptodate:
+            try:
+                self._request_coins_list()
+                self._request_marketcap_ranks()
+            except Exception as e:
+                if startup:
+                    raise OperationalException(
+                        f'Failed to download marketcap data from coingecko: {e}'
+                    )
+                else:
+                    self.log_once(
+                            'Failed to update marketcap data from coingecko: .'
+                            f'{e}. Using old data.',
+                            logger.warning)
+            else:
+                self._marketcap_cache['marketcaps_uptodate'] = True
```

Comment:
Well I think it is necessary, because if the cache is being invalidated you'd have to update the data from coingecko. If coingecko is not available for whatever reason, this will fail badly.
Storing the list in the class and just tracking allows ft to live with the old data until coingecko is back again.
I definitely do not want a hard dependency on something that is not the exchange. Using the TTLCache for that would introduce such a dependency. As far as I understand it the ttlcache code just drops the data as soon as the ttl is expired.

Reply:
Works for me :+1: